packages/framework/src/Model/Mail/EventListener/EnvelopeListener.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\FrameworkBundle\Model\Mail\EventListener;
+
+use Shopsys\FrameworkBundle\Model\Mail\MailerSettingProvider;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Mailer\Event\MessageEvent;
+use Symfony\Component\Mime\Address;
+use Symfony\Component\Mime\Header\MailboxListHeader;
+use Symfony\Component\Mime\Message;
+
+class EnvelopeListener implements EventSubscriberInterface
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Mail\MailerSettingProvider
+     */
+    protected MailerSettingProvider $mailerSettingProvider;
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Mail\MailerSettingProvider $mailerSettingProvider
+     */
+    public function __construct(MailerSettingProvider $mailerSettingProvider)
+    {
+        $this->mailerSettingProvider = $mailerSettingProvider;
+    }
+
+    /**
+     * @param \Symfony\Component\Mailer\Event\MessageEvent $event
+     */
+    public function onMessage(MessageEvent $event): void
+    {
+        if ($this->mailerSettingProvider->isMailerMasterEmailSet() === false) {
+            return;
+        }
+
+        $message = $event->getMessage();
+        if (!($message instanceof Message)) {
+            return;
+        }
+
+        $originalRecipients = [
+            ...$this->getAddressesFromMessageHeader($message, 'To'),
+            ...$this->getAddressesFromMessageHeader($message, 'Cc'),
+            ...$this->getAddressesFromMessageHeader($message, 'Bcc'),
+        ];
+        $allowedRecipients = $this->getAllowedRecipients($originalRecipients);
+
+        $event->getEnvelope()->setRecipients($allowedRecipients);
+    }
+
+    /**
+     * @param \Symfony\Component\Mime\Message $message
+     * @param string $headerName
+     * @return \Symfony\Component\Mime\Address[]
+     */
+    protected function getAddressesFromMessageHeader(Message $message, string $headerName): array
+    {
+        $header = $message->getHeaders()->get($headerName);
+        if ($header instanceof MailboxListHeader) {
+            return $header->getAddresses();
+        }
+
+        return [];
+    }
+
+    /**
+     * @param \Symfony\Component\Mime\Address[] $originalRecipients
+     * @return \Symfony\Component\Mime\Address[]
+     */
+    protected function getAllowedRecipients(array $originalRecipients): array
+    {
+        $allowedRecipients = [new Address($this->mailerSettingProvider->getMailerMasterEmailAddress())];
```

Comment:
PHP Stan on higher level would trigger error here. `getMailerMasterEmailAddress()` returns `string|null` and `Address` constructor expects `string`. So when `MAILER_MASTER_EMAIL_ADDRESS` env variable is not set, this would trigger fatal error.

Reply:
It's just a nitpick, as ENV variable should be set, but it an edge case that can occur.

Reply:
hm, that is a good catch, however, in the current implementation, the error would not occur even if the env variable was not set - `getAllowedRecipients` method is called within the public `onMessage` method only and in the public method, there is the following check at the beginning:
```
if ($this->mailerSettingProvider->isMailerMasterEmailSet() === false) {
   return;
}
```

I would say it is good enough but let's discuss it, please

Reply:
ok, I did some refactoring in 58c6686f8 - now it is ensured that only string is passed to the `Address` constructor, and an exception is raised when master mail is not set so PHPStan is satisfied even on level 9 :grin: I do not want to catch that exception as I actually do not know how the application should behave in such a situation - should the array of recipients be empty, or should we set at least the `$originalRecipients` array there? I do not have an answer for that, and the proper behavior should be guaranteed thanks to the `isMailerMasterEmailSet` check in `onMessage` method as mentioned above.

Please, let me know what you think about this solution :slightly_smiling_face: